File: q100-q199/q100-109/q103/q103.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (root == NULL) {
            return ans;
        }
        queue<TreeNode*> level_queue;
        level_queue.push(root);
        int level_cnt = 0;
        while (!level_queue.empty()) {
            ++level_cnt;
            int node_cnt = level_queue.size();
            vector<int> level;
            for (int i = 0; i < node_cnt; ++i) {
                TreeNode* node = level_queue.front();
                if (node->left != NULL) {
                    level_queue.push(node->left);
                }
                if (node->right != NULL) {
                    level_queue.push(node->right);
                }
                if (level_cnt % 2 != 0) {
                    level.push_back(node->val);
                } else {
                    level.insert(level.begin(), node->val);
                } 
                level_queue.pop();
            }
            ans.push_back(level);
        }
        return ans;
    }
};

File: q100-q199/q100-109/q103/q103.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import queue
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if root == None:
            return []
        ans = []
        level_cnt = 0
        level_queue = queue.SimpleQueue()
        level_queue.put(root)
        while not level_queue.empty():
            level_cnt += 1
            node_size = level_queue.qsize()
            level = []
            for i in range(0, node_size):
                node = level_queue.get()
                if node.left != None:
                    level_queue.put(node.left)
                if node.right != None:
                    level_queue.put(node.right)
                if level_cnt % 2 != 0:
                    level.append(node.val)
                else:
                    level.insert(0, node.val)
            ans.append(level)
        return ans
